lib/ruby_lsp/requests/document_highlight.rb

```diff
@@ -22,34 +22,49 @@ module Requests
     #   FOO # should be highlighted as "read"
     # end
     # ```
-    class DocumentHighlight < BaseRequest
+    class DocumentHighlight < Listener
       extend T::Sig
 
-      sig { params(document: Document, position: Document::PositionShape).void }
-      def initialize(document, position)
-        super(document)
+      ResponseType = type_member { { fixed: T::Array[Interface::DocumentHighlight] } }
 
-        @highlights = T.let([], T::Array[Interface::DocumentHighlight])
-        return unless document.parsed?
+      sig { override.returns(ResponseType) }
+      attr_reader :response
 
-        @target = T.let(find(position), T.nilable(Support::HighlightTarget))
+      sig do
+        params(
+          target: T.nilable(SyntaxTree::Node),
+          parent: T.nilable(SyntaxTree::Node),
+          emitter: EventEmitter,
+          message_queue: Thread::Queue,
+        ).void
       end
+      def initialize(target, parent, emitter, message_queue)
+        super(emitter, message_queue)
+
+        @response = T.let([], T::Array[Interface::DocumentHighlight])
+
+        return unless target && parent
 
-      sig { override.returns(T.all(T::Array[Interface::DocumentHighlight], Object)) }
-      def run
-        # no @target means the target is not highlightable
-        visit(@document.tree) if @document.parsed? && @target
-        @highlights
+        highlight_target =
+          case target
+          when *DIRECT_HIGHLIGHTS
+            Support::HighlightTarget.new(target)
+          when SyntaxTree::Ident
+            relevant_node = parent.is_a?(SyntaxTree::Params) ? target : parent
+            Support::HighlightTarget.new(relevant_node)
+          end
```

Comment:
Not a blocker, but just a thought: what if we move this to `Executor`? Would that maybe simplify things a bit?

Reply:
Consider the data type it uses (`HighlightTarget`) is still specific to this request, I prefer keeping it at the same place until we share it somewhere else.